File: python-leetcode/linked-list/19.删除链表的倒数第 N 个结点.py
from typing import Optional

from type import ListNode


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(-1, head)
        slow = fast = dummy

        # fast 先走 n 步
        for _ in range(n):
            assert fast is not None
            fast = fast.next

        # fast 再提前走一步，让 slow 指向待删节点的前驱
        assert fast is not None
        fast = fast.next

        # 同时移动
        while fast:
            assert slow is not None
            slow = slow.next
            fast = fast.next

        # 删除 slow.next
        assert slow is not None
        assert slow.next is not None
        slow.next = slow.next.next

        return dummy.next

File: go-leetcode/linked-list/19.删除链表的倒数第 N 个结点.go
package leetcode

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Val: -1, Next: head}
	slow, fast := dummy, dummy

	// fast 先走 n 步
	for i := 0; i < n; i++ {
		fast = fast.Next
	}

	// fast 再提前走一步，让 slow 指向待删节点的前驱
	fast = fast.Next

	// 同时移动
	for fast != nil {
		slow = slow.Next
		fast = fast.Next
	}

	// 删除 slow.Next
	slow.Next = slow.Next.Next

	return dummy.Next
}
